### ISSUE TITLE
Garden (1.19.14) fails to start on bosh-lite environment

### ISSUE THREAD

**Issue by kauana** [cloudfoundry/garden-runc-release#161]
## Description

The networking program has recently worked on a story that enables service discovery by default rather than using an [opsfile](https://github.com/cloudfoundry/cf-deployment/blob/master/operations/enable-service-discovery.yml) to do so.

However, when we created a [PR to cf-deployment](https://github.com/cloudfoundry/cf-deployment/pull/892) we learned that our changes were causing the RelInt's lite-deploy job to fail consistently ([check builds 2948-2950](https://release-integration.ci.cf-app.com/teams/main/pipelines/cf-deployment/jobs/lite-deploy/builds/2948)) and when our PR was reverted the job was green again.

It is worth noting that we made no changes to garden whatsoever and our changes did not cause a not bosh-lite environment to fail.

We have more context on the things we've tried in this [Slack thread](https://cloudfoundry.slack.com/archives/CFX13JK7B/p1594766757050600) which you are welcome to look at.

## Environment

* garden-runc-release version: 1.19.14
* Enabled configuration: n/a
* IaaS: GCP
* Stemcell version: ubuntu-xenial 621.76
* Kernel version: 4.15

## Steps to reproduce
**Given** a bosh-lite environment that uses the cf-deployment manifest with the service-discovery enabled by default changes from this [PR](https://github.com/cloudfoundry/cf-deployment/pull/892)
**Then** I see that Garden fails because it could not start

## Logs
From the garden logs when CI failed:
```json
/var/vcap/packages/greenskeeper/bin/system-preparation: line 26: echo: write error: Invalid argument
Could not permit all devices. This is normal in non systemd mode.
/var/vcap/packages/greenskeeper/bin/system-preparation: line 57: /sys/kernel/mm/transparent_hugepage/enabled: Read-only file system
Could not disable automatic transparent hugepage allocation. This is normal in bosh lite.
{"timestamp":"2020-07-15T18:54:02.161021993Z","level":"error","source":"grootfs","message":"grootfs.init-store.store-manager-init-store.mounting-fs-failed","data":{"allOpts":"loop,pquota,noatime","destination":"/var/vcap/data/grootfs/store/unprivileged","error":"exit status 32","session":"1.1","source":"/var/vcap/data/grootfs/store/unprivileged.backing-store","spec":{"UIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"GIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"StoreSizeBytes":51398832128},"storePath":"/var/vcap/data/grootfs/store/unprivileged"}}
{"timestamp":"2020-07-15T18:54:02.161469012Z","level":"info","source":"grootfs","message":"grootfs.init-store.store-manager-init-store.existing-backing-store-could-not-be-mounted: Mounting filesystem: exit status 32: mount: /var/vcap/data/grootfs/store/unprivileged.backing-store: failed to setup loop device: Operation not permitted\n","data":{"session":"1.1","spec":{"UIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"GIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"StoreSizeBytes":51398832128},"storePath":"/var/vcap/data/grootfs/store/unprivileged"}}
{"timestamp":"2020-07-15T18:54:02.163771245Z","level":"error","source":"grootfs","message":"grootfs.init-store.store-manager-init-store.overlayxfs-init-filesystem.mounting-fs-failed","data":{"allOpts":"remount,loop,pquota,noatime","destination":"/var/vcap/data/grootfs/store/unprivileged","error":"exit status 32","filesystemPath":"/var/vcap/data/grootfs/store/unprivileged.backing-store","session":"1.1.2","source":"/var/vcap/data/grootfs/store/unprivileged.backing-store","spec":{"UIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"GIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"StoreSizeBytes":51398832128},"storePath":"/var/vcap/data/grootfs/store/unprivileged"}}
{"timestamp":"2020-07-15T18:54:02.242143736Z","level":"error","source":"grootfs","message":"grootfs.init-store.store-manager-init-store.overlayxfs-init-filesystem.mounting-fs-failed","data":{"allOpts":"loop,pquota,noatime","destination":"/var/vcap/data/grootfs/store/unprivileged","error":"exit status 32","filesystemPath":"/var/vcap/data/grootfs/store/unprivileged.backing-store","session":"1.1.2","source":"/var/vcap/data/grootfs/store/unprivileged.backing-store","spec":{"UIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"GIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"StoreSizeBytes":51398832128},"storePath":"/var/vcap/data/grootfs/store/unprivileged"}}
{"timestamp":"2020-07-15T18:54:02.242322366Z","level":"error","source":"grootfs","message":"grootfs.init-store.store-manager-init-store.initializing-filesystem-failed","data":{"backingstoreFile":"/var/vcap/data/grootfs/store/unprivileged.backing-store","error":"Mounting filesystem: exit status 32: mount: /var/vcap/data/grootfs/store/unprivileged.backing-store: failed to setup loop device: Operation not permitted\n","session":"1.1","spec":{"UIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"GIDMappings":[{"HostID":4294967294,"NamespaceID":0,"Size":1},{"HostID":1,"NamespaceID":1,"Size":4294967293}],"StoreSizeBytes":51398832128},"storePath":"/var/vcap/data/grootfs/store/unprivileged"}}
{"timestamp":"2020-07-15T18:54:02.242364029Z","level":"error","source":"grootfs","message":"grootfs.init-store.init-store-failed","data":{"error":"initializing filesyztem: Mounting filesystem: exit status 32: mount: /var/vcap/data/grootfs/store/unprivileged.backing-store: failed to setup loop device: Operation not permitted\n","session":"1"}}
```

## Docker image (if relevant)
n/a

## Cause
The cause is not yet known.

## Similar Issue
Release Integration has encountered [this issue](https://cloudfoundry.slack.com/archives/C033RE5D6/p1555364813029600) before about a year ago, but unfortunately the issue was not created and it was not explored further :(

cc: @ameowlia @davewalter

**Comment by kieron-pivotal** [comment]
Hi @kauana ,

This is a really interesting problem!

Garden sets some restrictive permissions on devices for security reasons. When running garden in a diego-cell within bosh-lite, we have garden within garden. Garden needs to loop mount a file as an XFS filesystem for the container rootfs's, and the device restrictions imposed by the parent garden process would stop this. Therefore, on garden startup, it checks if it's running in a bosh-lite, and changes the device permissions for itself. However, it can only do this when there are no children of its cgroup. BPM creates child cgroups as it starts services. This means garden must be the first service to start before any BPM services.

There _seems_ to be an undocumented property of monit that we rely on. Monitrc files are created in /var/vcap/monit/job/ in reverse order of job inclusion in the instance group in the manifest. When monit starts, it checks the status of the last entry there, and attempts to start it if not running. It then starts the other files in ascending order. Very strange! But we need garden to be the last non-empty entry alphabetically in that directory.

The ops file uses
```
- type: replace
  path: /instance_groups/name=diego-cell/jobs/name=bosh-dns-adapter?
```
to insert the bosh-dns-adapter job, if it doesn't exist. And this just so happens to put it at the top of the jobs list, before cflinuxfs3-rootfs-setup (which is not a service) and garden, which means it starts first, not garden.

So, a workaround is to make sure bosh-dns-adapter is not inserted before garden. This will be fine if it's included directly in cf-deployment.yml. It might be trickier to get right in an ops file.

In the meantime, we will investigate if it's possible to move the device permission logic into a pre-start script to stop this being an issue, as relying on monit start order seems very fragile.

Kieron & @gcapizzi

**Comment by kieron-pivotal** [comment]
We've moved the device perms setting to a pre-start script and released garden-runc-release-v1.19.15. So no workarounds required now.

**Comment by ameowlia** [comment]
🔥 🔥 🔥 !

thank you and thanks for such a detailed description of the problem!
